### PR TITLE
MAINT: Factor publish to pypi workflow into dedicated file

### PR DESCRIPTION
**Pull request by choldgraf**
Currently we have a workflow embedded in `tests.yml` that is called only if a new release is made. However this shows us as "not run" in our test suite in github and messes with the passing tests condition.

This PR factors that out into a dedicated workflow file that is run on `publish` and _calls_ our test suite before running the publish command. That should mean the "Publish to PyPI" action never gets run unless we are publishing.

cc @AakashGfude this should make it easier to quickly see if the tests are happy in PRs

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+on:
+  release:
+    types: ["published"]
+
+jobs:
+  tests:
+    uses: ./.github/workflows/tests.yml
+  publish:
+    name: publish
+    needs: [tests] # require tests to pass before deploy runs
```
Comment (AakashGfude) [inline .github/workflows/publish.yml:14]: Do we also need 'docs-audit' here, or 'tests' is good enough you think?

**Review by AakashGfude — APPROVED** [review]
Thanks @choldgraf for this feature. Looks good to me. Apart from one comment above.

File: .github/workflows/publish.yml (hunk 1 above)
Comment (choldgraf) [reply .github/workflows/publish.yml:14]: This will run _all_ of the jobs that are in that workflow file, so it'll run `docs-audit` as well 👍

**Comment by choldgraf** [comment]
thanks for the approve @AakashGfude - I think those two unreported tests are bugs in github not due to these changes, so lemme merge and see if that resolves it

**Pull request merged**